Add many more types of VPC endpoints

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds many more types of VPC endpoints.

## 💭 Motivation and context ##

There is no need for our AWS API traffic to go back out to the internet, and it improves 
security if it does not.

See also:
- cisagov/cool-sharedservices-freeipa#63
- cisagov/cool-sharedservices-openvpn#61

## 🧪 Testing ##

All automated tests pass.  I also applied these changes to our COOL staging environment and verified that they function as expected.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

jsf9k committed:
Add many more types of VPC endpoints

dav3r commented on README.md:
```diff
@@ -125,8 +161,10 @@ At this point the `ProvisionNetworking` policy is attached to the
 
 | Name | Description |
 |------|-------------|
+| cloudwatch\_agent\_endpoint\_client\_security\_group | A security group for any instances that run the AWS CloudWatch agent.  This security groups allows such instances to communicate with the VPC endpoints that are required by the AWS CloudWatch agent. |
```
Typo:
```suggestion
| cloudwatch\_agent\_endpoint\_client\_security\_group | A security group for any instances that run the AWS CloudWatch agent.  This security group allows such instances to communicate with the VPC endpoints that are required by the AWS CloudWatch agent. |
```

dav3r commented on outputs.tf:
```diff
@@ -1,3 +1,8 @@
+output "cloudwatch_agent_endpoint_client_security_group" {
+  value       = aws_security_group.cloudwatch_agent_endpoint_client
+  description = "A security group for any instances that run the AWS CloudWatch agent.  This security groups allows such instances to communicate with the VPC endpoints that are required by the AWS CloudWatch agent."
```
Typo:
```suggestion
  description = "A security group for any instances that run the AWS CloudWatch agent.  This security group allows such instances to communicate with the VPC endpoints that are required by the AWS CloudWatch agent."
```

dav3r commented on vpc_endpoints.tf:
```diff
@@ -2,6 +2,186 @@
+
+  private_dns_enabled = true
+  security_group_ids = [
+    # The CloudWatch agent reads a few pieces of data from the ec2
```
Capitalization:
```suggestion
    # The CloudWatch agent reads a few pieces of data from the EC2
```

dav3r reviewed (approved):
Looks great - I could only find a couple of superficial quibbles.

jsf9k committed:
Correct grammar

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Correct capitalization

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj commented on vpc_endpoints.tf:
```diff
@@ -2,6 +2,186 @@
+# Note that SSM requires several other endpoints to function properly.
+# See here for more details:
+# https://docs.aws.amazon.com/systems-manager/latest/userguide/setup-create-vpc.html#sysman-setting-up-vpc-create
+resource "aws_vpc_endpoint_subnet_association" "ssm" {
+  provider = aws.sharedservicesprovisionaccount
+
+  for_each = toset(var.private_subnet_cidr_blocks)
+
+  subnet_id       = module.private.subnets[each.value].id
+  vpc_endpoint_id = aws_vpc_endpoint.ssm.id
+}
+resource "aws_vpc_endpoint_subnet_association" "ec2" {
+  provider = aws.sharedservicesprovisionaccount
+
+  for_each = toset(var.private_subnet_cidr_blocks)
+
+  subnet_id       = module.private.subnets[each.value].id
+  vpc_endpoint_id = aws_vpc_endpoint.ec2.id
+}
+resource "aws_vpc_endpoint_subnet_association" "ec2messages" {
+  provider = aws.sharedservicesprovisionaccount
+
+  for_each = toset(var.private_subnet_cidr_blocks)
+
+  subnet_id       = module.private.subnets[each.value].id
+  vpc_endpoint_id = aws_vpc_endpoint.ec2messages.id
+}
+resource "aws_vpc_endpoint_subnet_association" "kms" {
+  provider = aws.sharedservicesprovisionaccount
+
+  for_each = toset(var.private_subnet_cidr_blocks)
+
+  subnet_id       = module.private.subnets[each.value].id
+  vpc_endpoint_id = aws_vpc_endpoint.kms.id
+}
```
Unless there's a specific reason for this ordering I think it should mirror the order of the `aws_vpc_endpoint` definitions.

```suggestion
resource "aws_vpc_endpoint_subnet_association" "ec2" {
  provider = aws.sharedservicesprovisionaccount

  for_each = toset(var.private_subnet_cidr_blocks)

  subnet_id       = module.private.subnets[each.value].id
  vpc_endpoint_id = aws_vpc_endpoint.ec2.id
}
resource "aws_vpc_endpoint_subnet_association" "ec2messages" {
  provider = aws.sharedservicesprovisionaccount

  for_each = toset(var.private_subnet_cidr_blocks)

  subnet_id       = module.private.subnets[each.value].id
  vpc_endpoint_id = aws_vpc_endpoint.ec2messages.id
}
resource "aws_vpc_endpoint_subnet_association" "kms" {
  provider = aws.sharedservicesprovisionaccount

  for_each = toset(var.private_subnet_cidr_blocks)

  subnet_id       = module.private.subnets[each.value].id
  vpc_endpoint_id = aws_vpc_endpoint.kms.id
}
resource "aws_vpc_endpoint_subnet_association" "ssm" {
  provider = aws.sharedservicesprovisionaccount

  for_each = toset(var.private_subnet_cidr_blocks)

  subnet_id       = module.private.subnets[each.value].id
  vpc_endpoint_id = aws_vpc_endpoint.ssm.id
}
```

mcdonnnj replied:
@jsf9k I see you resolved this but the change was not committed and there's no explanatory comment for why the change should not be implemented. I have unresolved it so that either of the two outcomes mentioned above can occur.

jsf9k replied:
Apologies, I don't know what happened there.

mcdonnnj commented on ssm_endpoint_sg.tf:
```diff
@@ -0,0 +1,23 @@
+  vpc_id = aws_vpc.the_vpc.id
+
+  tags = {
+    Name = "SSM endpoints"
```
Is there a reason the resource name is singular but the tag name is plural?

jsf9k replied:
That happened because there is a single VPC endpoint service that is attached to multiple subnets, each in its own availability zone.  I can change the `tags` to use the singular.

jsf9k replied:
Please see commit fd19f41.

mcdonnnj commented on ssm_agent_endpoint_sg.tf:
```diff
@@ -0,0 +1,23 @@
+  vpc_id = aws_vpc.the_vpc.id
+
+  tags = {
+    Name = "SSM agent endpoints"
```
Is there a reason the resource name is singular but the tag name is plural?

jsf9k replied:
The security group is applied to multiple VPC endpoints, since access to multiple VPC endpoints is required to make use of the SSM Agent.

mcdonnnj commented on sts_endpoint_client_sg.tf:
```diff
@@ -0,0 +1,22 @@
+  }
+}
+
+# Allow egress via HTTPS from the STS endpoint security group.
```
```suggestion
# Allow egress via HTTPS to the STS endpoint security group.
```

mcdonnnj commented on cloudwatch_agent_endpoint_sg.tf:
```diff
@@ -0,0 +1,24 @@
+  vpc_id = aws_vpc.the_vpc.id
+
+  tags = {
+    Name = "CloudWatch agent endpoints"
```
Is there a reason the resource name is singular but the tag name is plural?

jsf9k replied:
The security group is applied to multiple VPC endpoints, since access to multiple VPC endpoints is required to make use of the CloudWatch Agent.

mcdonnnj reviewed (commented):
This makes sense. I have the same question repeated a few times and a couple of suggestions.

jsf9k committed:
Correct preposition

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct number of word in tag

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on ssm_endpoint_sg.tf:
```diff
@@ -0,0 +1,23 @@
+  vpc_id = aws_vpc.the_vpc.id
+
+  tags = {
+    Name = "SSM endpoint"
```
So I know I questioned you about this previously and you changed it to singular. However upon revisiting this and checking against `vpc_endpoints.tf` I see that this security group is applied to multiple VPC endpoints. I believe this should be plural to be consistent with the other SGs applied to multiple VPC endpoints but please let me know if I'm reading this incorrectly.

```suggestion
    Name = "SSM endpoints"
```

jsf9k replied:
Ugh.  Please see commit fa1e368.

mcdonnnj replied:
Apologies and thank you for your patience 😅

mcdonnnj reviewed (approved):
Strong work 💪💪💪 I have one comment for consideration but this is otherwise looking good.

jsf9k committed:
Revert "Correct number of word in tag"

This reverts commit bd2765412b5f0a4140691f4a4cb0614940ad33c0.

@mcdonnnj has decided that the name tag should be plural after all.

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Reorder declaration of endpoint subnet associations

The new order is consistent with the order of declaration of the VPC endpoints themselves, and is also alphabetical.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
